### ISSUE TITLE
Linking problems when building dll

### ISSUE THREAD

**Issue by abeham** [igraph/igraph#956]
I managed to compile igraph master branch as a static library after merging the pull request (#955). However, I still fail at compiling igraph as a dynamic link library (dll) or compiling a c++ dll where I link igraph.lib statically. I get linker errors, e.g.

> Error	931	error LNK2019: unresolved external symbol _igraph_2wheap_init referenced in function _igraph_i_betweenness_estimate_weighted	somepath\igraph\igraph-0.8.0-pre+38f0c2d-msvc\centrality.obj	igraph
> Error	932	error LNK2001: unresolved external symbol _igraph_2wheap_init	somepath\igraph\igraph-0.8.0-pre+38f0c2d-msvc\community.obj	igraph

I looked up the possible causes for this error at MSDN ([LNK2001] and [LNK2019]), but couldn't find a solution. igraph_2wheap_init is declared in igraph_internal_types.h and implemented in heap.c. Both are present in the project and compiled.

I only managed to resolve one linker error by declaring MSDOS in the preprocessor (related to s_paus.c)
I used VS2013.

[LNK2001]: https://msdn.microsoft.com/en-us/library/f6xx1b1z.aspx
[LNK2019]: https://msdn.microsoft.com/en-us/library/799kze2z.aspx

**Comment by ntamas** [comment]
Summon @janschulz - have you seen something like this before?

**Comment by abeham** [comment]
I think I found the problem with respect to the igraph_2wheap_* linker errors. There are two files heap.c and heap.cc which seem to overwrite each others' obj files (see this [stackoverflow answer]). Thus the symbols get lost. I renamed heap.c to heap2.c and reduced 69 linker errors down to 18. A similar problem with graph.c and graph.cc -> down to 15 errors.

The remaining ones are like
> Error	4	error LNK2019: unresolved external symbol _igraphdtrsm referenced in function _r_cholmod_super_numeric	somepath\igraph-0.8.0-pre+38f0c2d-msvc\cholmod_super_numeric.obj	igraph
> Error	5	error LNK2001: unresolved external symbol _igraphdtrsm	somepath\igraph-0.8.0-pre+38f0c2d-msvc\cholmod_super_solve.obj	igraph
> Error	21	error LNK2019: unresolved external symbol "public: __thiscall prpack::prpack_base_graph::~prpack_base_graph(void)" (??1prpack_base_graph@prpack@@QAE@XZ) referenced in function "public: __thiscall prpack::prpack_igraph_graph::~prpack_igraph_graph(void)" (??1prpack_igraph_graph@prpack@@QAE@XZ)	C:\Users\P40311\Documents\GitHub\igraph\igraph-0.8.0-pre+38f0c2d-msvc\prpack.obj	igraph

Regarding the first errors, I saw that in cholmod_blas.h in line 55 one would define BLAS_NO_UNDERSCORE for Windows when not compiled by MINGW32 or CYGWIN32. The functions are however defined with underscore so if I remove that definition, cholmod links fine. Alternatively one could probably write LAPACK_DPOTRF instead of igraphdpotrf_ as function name in e.g. dpotrf.c. I don't know which is more feasible.

That brings the errors down to 7

I saw that only prpack.cpp was included in the source files of the project, but not the other files from src\prpack. I assume they don't get compiled then. I included the cpp files from src\prpack in the project's sources and voilà I do have an igraph.dll built with VS2013. I am a bit suspicious still as the dll is just 374kb in size. This seems to be very small.

So in summary the changes:
 1. Rename heap.c to heap2.c and graph.c to graph2.c
 2. Uncomment the definition of BLAS_NO_UNDERSCORE in cholmod_blas.h
 3. Include src\prpack\*.cpp into the project's sources

[stackoverflow hint]: http://stackoverflow.com/a/12826901/189767

**Comment by abeham** [comment]
I still want to add __declspec(dllexport) respectively__declspec(dllimport) declarations to several of the igraph functions in order to be able to p/invoke into the dll from e.g. C#. I thought about doing this with a macro (as is recommended on msdn)

```
#if defined (WIN32) || defined (_WIN32) || defined (WIN64) || defined (_WIN64)
# if defined (__MINGW32__) || defined (__CYGWIN32__)
#  define DECLDIR /**/
# else
#  ifdef IGRAPH_EXPORTS
#   define DECLDIR __declspec(dllexport)
#  else
#   define DECLDIR __declspec(dllimport)
#  endif
# endif
#else
# define DECLDIR /**/
#endif
```

And then using the directive like this in the header files

`DECLDIR int igraph_emtpy(igraph_t *graph, igraph_integer_t v, igraph_bool_t directed);`

Where should I put the preprocessor statements that define DECLDIR? Is there a header file that gets included first?

**Comment by ntamas** [comment]
FIrst of all, thanks for your hard work on this! I will take some time to merge your suggestions into the master branch once you tell me that you have successfully built an igraph dll that is suitable for your purposes.

There is no header file that reliably gets included "first" from all other header files, although at this point it might make sense to create one anyway. You can see that all the header files in `include/` contain a boilerplate like this:

```c
#undef __BEGIN_DECLS
#undef __END_DECLS
#ifdef __cplusplus
# define __BEGIN_DECLS extern "C" {
# define __END_DECLS }
#else
# define __BEGIN_DECLS /* empty */
# define __END_DECLS /* empty */
#endif
```

This is to make the headers work both in C, C++ and mixed-C-and-C++ projects. I can easily imagine moving this to an `igraph_decls.h` header and including that header from every other header. Then your `DECLDIR` macro could also go into `igraph_decls.h`.

**Comment by abeham** [comment]
Yes that would be a good idea. I already have a fork and will include that. I hope that I can send you a pull request (still a novice wrt git). The DLL I have is up and running in x86 and x64 and I can call into it from C# (still have to commit some changes). I am currently writing a C# wrapper so that I abstract igraph functions and plan to include that within HeuristicLab.

I saw that some files missed these directives. On purpose or oversight?

- igraph_cohesive_blocks.h
- igraph_estack.h
- igraph_lsap.h
- igraph_marked_queue.h
- igraph_scg.h
- igraph_sparsemat.h (there is igraph_spmatrix.h though)

and some others for which it's probably okay (pmt and internal)

- igraph_arpack_internal.h
- igraph*_pmt.h
- igraph_blas_internal.h
- igraph_cliquer.h
- igraph_flow_internal.h
- igraph_glpk_support.h
- igraph_hacks_internal.h
- igraph_lapack_internal.h

**Comment by abeham** [comment]
Okay, I think I finished most changes and created a pull request.

**Comment by ntamas** [comment]
Continuing discussion in #958
